client/lib/wp/sync-handler/cache.js

```diff
@@ -0,0 +1,178 @@
+/**
+ * External dependencies
+ */
+import moment from 'moment';
+import debugFactory from 'debug';
+import ms from 'ms';
+
+/**
+ * Internal dependencies
+ */
+import warn from 'lib/warn';
+import { getLocalForage } from 'lib/localforage';
+
+/**
+ * Module variables
+ */
+const localforage = getLocalForage();
+const debug = debugFactory( 'calypso:sync-handler:cache' );
+
+const RECORDS_LIST_KEY = 'records-list';
+const SYNC_RECORD_REGEX = /^sync-record-\w+$/;
+const LIFETIME = '2 days';
+
+/**
+ * Check it the given key is a `sync-record` key
+ *
+ * @param {String} key - record key
+ * @return {Boolean} `true` if it's a sync-record-<key>
+ */
+const isSyncRecordKey = key => {
+	const isSyncRecord = SYNC_RECORD_REGEX.test( key );
+
+	if ( isSyncRecord ) {
+		debug( '%o is a sync-record', key );
+	}
+	return isSyncRecord;
+}
+
+export class Cache {
+	getAll() {
+		return localforage.getItem( RECORDS_LIST_KEY );
+	}
+
+	/**
+	 * Add the given `key` into the records-list object
+	 * adding at the same a marktime (now).
+	 * If the pair key-mark already exists it will be updated.
+	 *
+	 * @param {String} key - record key
+	 * @return {Promise} promise
+	 */
+	addItem( key ) {
+		return this.filterByKey( key ).then( records => {
+			records.unshift( { key, mark: Date.now() } );
+			debug( 'adding %o to list (%s)', key, records.length );
+			return localforage.setItem( RECORDS_LIST_KEY, records );
+		} );
+	}
+
+	removeItem( key ) {
+		return this.filterByKey( key ).then( records => {
+			return localforage.setItem( RECORDS_LIST_KEY, records );
+		} );
+	}
+
+	/**
+	 * Retrieve all records filter by the given key
+	 *
+	 * @param {String} key - compare records with this key
+	 * @return {Promise} promise
+	 */
+	filterByKey( key ) {
+		return new Promise( ( resolve, reject ) => {
+			this.getAll().then( records => {
+				if ( ! records || ! records.length ) {
+					debug( 'No records stored' );
+					return resolve( [] );
+				}
+
+				// filter records by the given key
+				records = records.filter( item => {
+					if ( item.key === key ) {
+						debug( '%o exists. Removing ...', key );
+					}
+					return item.key !== key;
+				} );
+
+				resolve( records );
+			} ).catch( reject );
```

Comment:
We should be able to coerce the output of `getAll()` into an array and skip a bunch of error checking and make the flow a bit clearer.

Also, `filter` is a pretty standard term meaning to pull out all the elements from a list that match a certain pattern, so I would expect this function to return a list of notes that match the key, but it appears like it returns the opposite - a list of the elements that _don't_ match that key. I recommend that we change this naming to something less confusing, such as `dropByKey()` or `dropWithKey()`

```js
import castArray from 'lodash/lang/castArray';

dropByKey( key ) {
	const dropMatches = records =>
		records.filter( negate( matchesProperty( 'key', key ) ) );

	return new Promise( ( resolve, reject ) => {
		this.getAll()
			.then( castArray )
			.then( dropMatches )
			.then( resolve )
			.catch( reject );
	} );
}
```